File: src/kx.js
// Compact werkkzeug3 player-export reader. The byte order and compact scalar
// encodings mirror KDoc::Init in Farbrausch's released source.
import { OPC_BLOB, OPC_FLEXINPUT } from './abi.js';

const MAX_OP_ROOT = 16;
// The serialized convention has an eight-bit data-word count. Debris contains
// two intentional one-word packing overruns (18 fields for 17 data words), so
// 256 is both sufficient for the exact party data and a useful malformed-input
// ceiling before a packing string is multiplied by thousands of operations.
const MAX_PACKING_LENGTH = 256;
const MAX_DECODED_PARAMETER_SLOTS = 4 * 1024 * 1024;

const ANIM_PAYLOAD_SIZES = new Uint8Array([
  0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 16, 4, 4, 2, 0, 0, 0, 8, 2, 0,
  0, 0,
]);

const floatScratch = new DataView(new ArrayBuffer(4));

function bitsToFloat(bits) {
  floatScratch.setUint32(0, bits >>> 0, true);
  return floatScratch.getFloat32(0, true);
}

function align4(value) {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(`invalid byte length ${value}`);
  }
  const remainder = value % 4;
  return remainder ? value + 4 - remainder : value;
}

function inputCount(convention) {
  return (convention & 0x00000f00) >>> 8;
}

function linkCount(convention) {
  return (convention & 0x0000f000) >>> 12;
}

function stringCount(convention) {
  return (convention & 0x00070000) >>> 16;
}

function splineCount(convention) {
  return (convention & 0x00700000) >>> 20;
}

class Reader {
  constructor(bytes, offset = 0) {
    if (!Number.isSafeInteger(offset) || offset < 0 || offset > bytes.byteLength) {
      throw new RangeError(`invalid KX offset ${offset}`);
    }
    this.bytes = bytes;
    this.view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    this.pos = offset;
  }

  require(size, what = 'data') {
    if (!Number.isSafeInteger(size) || size < 0 || size > this.bytes.byteLength - this.pos) {
      throw new Error(
        `truncated ${what} at 0x${this.pos.toString(16)}: need ${size} bytes`,
      );
    }
  }

  skip(size, what = 'data') {
    this.require(size, what);
    const start = this.pos;
    this.pos += size;
    return start;
  }

  u8() {
    this.require(1);
    return this.bytes[this.pos++];
  }

  u16() {
    this.require(2);
    const value = this.view.getUint16(this.pos, true);
    this.pos += 2;
    return value;
  }

  s16() {
    this.require(2);
    const value = this.view.getInt16(this.pos, true);
    this.pos += 2;
    return value;
  }

  u24() {
    const a = this.u8();
    const b = this.u8();
    const c = this.u8();
    return a | (b << 8) | (c << 16);
  }

  u32() {
    this.require(4);
    const value = this.view.getUint32(this.pos, true);
    this.pos += 4;
    return value;
  }

  s32() {
    this.require(4);
    const value = this.view.getInt32(this.pos, true);
    this.pos += 4;
    return value;
  }

  compactShort() {
    let value = this.u8();
    if (value & 0x80) value = (value & 0x7f) | (this.u8() << 7);
    return value;
  }

  string(what = 'string', maxLength = Infinity) {
    const start = this.pos;
    while (this.pos < this.bytes.byteLength && this.bytes[this.pos] !== 0) {
      if (this.pos - start >= maxLength) {
        throw new Error(`${what} at 0x${start.toString(16)} exceeds ${maxLength} bytes`);
      }
      this.pos++;
    }
    if (this.pos === this.bytes.byteLength) {
      throw new Error(`unterminated ${what} at 0x${start.toString(16)}`);
    }
    let result = '';
    // Avoid one character-at-a-time concatenation and an unbounded spread call
    // for malformed long operator strings while preserving byte-for-code-unit
    // semantics (the native format is not UTF-8).
    for (let i = start; i < this.pos; i += 0x4000) {
      result += String.fromCharCode(...this.bytes.subarray(i, Math.min(this.pos, i + 0x4000)));
    }
    this.pos++;
    return result;
  }

  f16() {
    const first = this.u8();
    if (first === 0x00) return 0;
    if (first === 0x80) return 1;
    if (first === 0x01) return 0.5;
    if (first === 0x81) return 0.25; // exact released-reader behavior
    const value = (first << 8) | this.u8();
    const bits =
      ((value & 0x8000) << 16) |
      (((((value >>> 10) & 31) + 128 - 16) & 0xff) << 23) |
      ((value & 1023) << 13);
    return bitsToFloat(bits);
  }

  f24() {
    const first = this.u8();
    if (first === 0x00) return 0;
    if (first === 0x01) return 1;
    if (first === 0xff) return -1;
    const second = this.u16();
    return bitsToFloat(
      (first << 23) | ((second & 0x8000) << 16) | ((second & 0x7fff) << 8),
    );
  }
}

function parameter(reader, packing) {
  switch (packing.toLowerCase()) {
    case 'g': return reader.f16();
    case 'f': return reader.f24();
    case 'e': return reader.s16() / 4096;
    case 'i': return reader.s32();
    case 's': return reader.s16();
    case 'b': return reader.u8();
    case 'c': return reader.u32();
    case 'm': return reader.u24();
    default: return null;
  }
}

function animation(reader) {
  const start = reader.pos;
  for (;;) {
    const command = reader.u8();
    if (command >= 0x80) {
      reader.skip(1, 'animation store operand');
    } else {
      if (command >= ANIM_PAYLOAD_SIZES.length) {
        throw new Error(`unknown animation opcode 0x${command.toString(16)}`);
      }
      reader.skip(ANIM_PAYLOAD_SIZES[command], 'animation operand');
    }
    if (command === 1) return reader.bytes.subarray(start, reader.pos);
  }
}

function asBytes(input) {
  if (input instanceof Uint8Array) return input;
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (ArrayBuffer.isView(input)) {
    return new Uint8Array(input.buffer, input.byteOffset, input.byteLength);
  }
  throw new TypeError('parseKX expects an ArrayBuffer or typed-array view');
}

function parseKX(input, offset = 0) {
  const bytes = asBytes(input);
  const reader = new Reader(bytes, offset);
  const start = offset;
  const sections = { header: 0 };

  const flags = reader.u32();
  if (flags & ~7) throw new Error(`unsupported KX flags 0x${flags.toString(16)}`);

  let name = null;
  if (flags & 1) {
    const nameStart = reader.pos;
    reader.skip(32, 'demo name');
    let length = 0;
    while (length < 32 && bytes[nameStart + length] !== 0) length++;
    name = String.fromCharCode(...bytes.subarray(nameStart, nameStart + length));
  }

  const songSize = reader.u32();
  const songOffset = reader.skip(align4(songSize), 'embedded song');
  sections.song = songOffset - start;

  let sampleOffset = null;
  let sampleSize = 0;
  if (flags & 2) {
    sampleSize = reader.u32();
    sampleOffset = reader.skip(align4(sampleSize), 'embedded samples');
    sections.samples = sampleOffset - start;
  }

  sections.document = reader.pos - start;
  const songBPMFixed = reader.u32();
  const songLength = reader.u32();
  const operationCount = reader.compactShort();
  const documentSplineCount = reader.compactShort();
  const roots = Array.from({ length: MAX_OP_ROOT }, () => reader.compactShort());

  const classes = [];
  for (;;) {
    const convention = reader.u32();
    if (convention === 0) break;
    if (classes.length === 128) throw new Error('more than 128 operator classes');
    classes.push({
      convention,
      id: reader.u16(),
      packing: reader.string('packing', MAX_PACKING_LENGTH),
    });
  }

  sections.graph = reader.pos - start;
  const operations = [];
  for (let opIndex = 0; opIndex < operationCount; opIndex++) {
    const typeByte = reader.u8();
    const classIndex = typeByte & 0x7f;
    const opClass = classes[classIndex];
    if (!opClass) throw new Error(`operator ${opIndex} has missing class ${classIndex}`);
    let count;
    if (typeByte & 0x80) count = 1;
    else if (opClass.convention & OPC_FLEXINPUT) count = reader.u8();
    else count = inputCount(opClass.convention);

    const inputs = [];
    for (let i = 0; i < count; i++) {
      const delta = typeByte & 0x80 ? 0 : reader.compactShort();
      const target = opIndex - 1 - delta;
      if (target < 0) throw new Error(`operator ${opIndex} has invalid input delta ${delta}`);
      inputs.push(target);
    }
    operations.push({
      id: opIndex,
      classIndex,
      classId: opClass.id,
      inputs,
      links: [],
      parameters: [],
      strings: [],
      splines: [],
      blobSize: 0,
      blob: null,
      animation: null,
    });
  }

  // Parameters are decoded into ordinary JavaScript arrays. A malicious class
  // table made entirely of no-byte '-' fields could otherwise amplify a small
  // graph into millions of array slots before the reader reaches another
  // bounds check. Real byte-consuming fields naturally stay below this ratio;
  // the exact party document uses 142,600 slots for 490,077 input bytes.
  const classOperationCounts = new Uint32Array(classes.length);
  for (const op of operations) classOperationCounts[op.classIndex]++;
  let decodedParameterSlots = 0;
  for (let classIndex = 0; classIndex < classes.length; classIndex++) {
    decodedParameterSlots += classOperationCounts[classIndex] * classes[classIndex].packing.length;
  }
  const parameterSlotBudget = Math.min(
    MAX_DECODED_PARAMETER_SLOTS,
    Math.max(4096, bytes.byteLength * 4),
  );
  if (decodedParameterSlots > parameterSlotBudget) {
    throw new Error(
      `decoded parameter table is too large: ${decodedParameterSlots} slots ` +
      `(limit ${parameterSlotBudget})`,
    );
  }

  for (const op of operations) {
    const count = linkCount(classes[op.classIndex].convention);
    for (let i = 0; i < count; i++) {
      const encoded = reader.compactShort();
      const target = encoded ? encoded - 1 : null;
      if (target !== null && target >= operationCount) throw new Error('invalid link target');
      op.links.push(target);
    }
  }

  sections.parameters = reader.pos - start;
  for (let classIndex = 0; classIndex < classes.length; classIndex++) {
    const opClass = classes[classIndex];
    for (const op of operations) {
      if (op.classIndex !== classIndex) continue;
      op.parameters = Array.from(opClass.packing, kind => parameter(reader, kind));
      const strings = stringCount(opClass.convention);
      for (let i = 0; i < strings; i++) op.strings.push(reader.string('operator string'));
      const splines = splineCount(opClass.convention);
      for (let i = 0; i < splines; i++) {
        const reference = reader.u16();
        const target = reference ? reference - 1 : null;
        if (target !== null && target >= documentSplineCount) {
          throw new Error(`operator ${op.id} has invalid spline target ${target}`);
        }
        op.splines.push(target);
      }
      if (opClass.convention & OPC_BLOB) op.blobSize = reader.u32();
    }
  }

  sections.animations = reader.pos - start;
  for (const op of operations) op.animation = animation(reader);

  sections.events = reader.pos - start;
  const eventCount = reader.compactShort();
  const events = [];
  for (let i = 0; i < eventCount; i++) {
    const operation = reader.compactShort();
    if (operation >= operationCount) throw new Error('event refers outside graph');
    const event = {
      operation,
      start: reader.s32(),
      end: reader.s32(),
      velocity: reader.f24(),
      modulation: reader.f24(),
      select: reader.u8(),
      scale: [reader.f24(), reader.f24(), reader.f24()],
      rotate: [reader.f24(), reader.f24(), reader.f24()],
      translate: [reader.f24(), reader.f24(), reader.f24()],
      color: reader.u32(),
      spline: (() => { const ref = reader.compactShort(); return ref ? ref - 1 : null; })(),
      startInterval: reader.f16(),
      endInterval: reader.f16(),
      flags: reader.u8(),
    };
    if (event.spline !== null && event.spline >= documentSplineCount) {
      throw new Error(`event ${i} has invalid spline target ${event.spline}`);
    }
    events.push(event);
  }

  sections.splines = reader.pos - start;
  const splines = [];
  for (let i = 0; i < documentSplineCount; i++) {
    const descriptor = reader.u8();
    const count = descriptor & 7;
    if (count > 4) throw new Error(`spline ${i} has ${count} channels`);
    const channels = [];
    for (let channel = 0; channel < count; channel++) {
      const keys = [];
      const keyCount = reader.compactShort();
      for (let key = 0; key < keyCount; key++) {
        keys.push({ time: reader.u8() / 255, value: reader.f16() });
      }
      channels.push(keys);
    }
    splines.push({ interpolation: descriptor >>> 3, channels });
  }

  sections.blobs = reader.pos - start;
  for (const op of operations) {
    if (!op.blobSize) continue;
    const blobOffset = reader.skip(op.blobSize, 'operator blob');
    op.blob = bytes.subarray(blobOffset, blobOffset + op.blobSize);
  }
  sections.end = reader.pos - start;

  return {
    bytes: bytes.subarray(start, reader.pos),
    flags,
    buzzTiming: Boolean(flags & 4),
    name,
    song: bytes.subarray(songOffset, songOffset + songSize),
    samples: sampleOffset === null
      ? new Uint8Array()
      : bytes.subarray(sampleOffset, sampleOffset + sampleSize),
    songBPMFixed,
    songBPM: songBPMFixed / 65536,
    songLength,
    roots,
    classes,
    operations,
    events,
    splines,
    sections,
  };
}

function summarizeKX(document) {
  const classCounts = new Uint32Array(document.classes.length);
  let animationBytes = 0;
  let blobBytes = 0;
  let blobCount = 0;
  for (const op of document.operations) {
    classCounts[op.classIndex]++;
    animationBytes += op.animation.byteLength;
    if (op.blobSize) {
      blobCount++;
      blobBytes += op.blobSize;
    }
  }
  return {
    size: document.bytes.byteLength,
    flags: document.flags,
    name: document.name,
    buzzTiming: document.buzzTiming,
    songSize: document.song.byteLength,
    sampleSize: document.samples.byteLength,
    songBPM: document.songBPM,
    songLength: document.songLength,
    operationCount: document.operations.length,
    eventCount: document.events.length,
    splineCount: document.splines.length,
    animationBytes,
    blobCount,
    blobBytes,
    sections: document.sections,
    classes: document.classes.map((opClass, index) => ({
      ...opClass,
      operationCount: classCounts[index],
    })),
  };
}

export { MAX_OP_ROOT, OPC_BLOB, OPC_FLEXINPUT, parseKX, summarizeKX };

File: tools/kx.py
#!/usr/bin/env python3
"""Read the compact werkkzeug3 player export used by fr-041: debris.

The layout is a direct, bounds-checked transcription of ``KDoc::Init`` from
Farbrausch's released werkkzeug3 source.  It deliberately does not depend on
the editor project format or any Windows code, which makes it suitable both
for extracting the party build's embedded document and for producing fixtures
for the browser reader.
"""

from __future__ import annotations

import hashlib
import json
import math
import struct
from dataclasses import dataclass
from pathlib import Path
from typing import Any


MAX_OP_ROOT = 16
OPC_FLEXINPUT = 0x80000000
OPC_BLOB = 0x00080000

# Payload bytes following the command byte. This is the CmdSize table in
# kdoc.cpp. KA_LOG2/KA_POW2 were added after the table and have no operands.
ANIM_PAYLOAD_SIZES = (
    0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 16, 4, 4, 2, 0, 0, 0, 8, 2, 0,
    0, 0,
)
KA_END = 0x01


class KxFormatError(ValueError):
    """The byte stream cannot be a valid compact KX document."""


class Cursor:
    def __init__(self, data: bytes | memoryview, offset: int = 0) -> None:
        self.data = memoryview(data)
        self.pos = offset

    def require(self, size: int, what: str = "data") -> None:
        if size < 0 or self.pos + size > len(self.data):
            raise KxFormatError(
                f"truncated {what} at {self.pos:#x}: need {size:#x} bytes, "
                f"have {max(0, len(self.data) - self.pos):#x}"
            )

    def take(self, size: int, what: str = "data") -> bytes:
        self.require(size, what)
        start = self.pos
        self.pos += size
        return bytes(self.data[start : self.pos])

    def skip(self, size: int, what: str = "data") -> int:
        self.require(size, what)
        start = self.pos
        self.pos += size
        return start

    def u8(self) -> int:
        self.require(1)
        value = self.data[self.pos]
        self.pos += 1
        return value

    def u16(self) -> int:
        return struct.unpack("<H", self.take(2))[0]

    def s16(self) -> int:
        return struct.unpack("<h", self.take(2))[0]

    def u24(self) -> int:
        raw = self.take(3)
        return raw[0] | (raw[1] << 8) | (raw[2] << 16)

    def u32(self) -> int:
        return struct.unpack("<I", self.take(4))[0]

    def s32(self) -> int:
        return struct.unpack("<i", self.take(4))[0]

    def compact_short(self) -> int:
        value = self.u8()
        if value & 0x80:
            value = (value & 0x7F) | (self.u8() << 7)
        return value

    def cstring(self, what: str = "string") -> str:
        start = self.pos
        while self.pos < len(self.data) and self.data[self.pos] != 0:
            self.pos += 1
        if self.pos == len(self.data):
            raise KxFormatError(f"unterminated {what} at {start:#x}")
        raw = bytes(self.data[start : self.pos])
        self.pos += 1
        return raw.decode("latin-1")

    def f16(self) -> float:
        first = self.u8()
        if first == 0x00:
            return 0.0
        if first == 0x80:
            return 1.0
        if first == 0x01:
            return 0.5
        if first == 0x81:
            # This surprising value is what the released reader returns.
            return 0.25
        value = (first << 8) | self.u8()
        bits = (
            ((value & 0x8000) << 16)
            | (((((value >> 10) & 31) + 128 - 16) & 0xFF) << 23)
            | ((value & 1023) << 13)
        )
        return struct.unpack("<f", struct.pack("<I", bits))[0]

    def f24(self) -> float:
        first = self.u8()
        if first == 0x00:
            return 0.0
        if first == 0x01:
            return 1.0
        if first == 0xFF:
            return -1.0
        second = self.u16()
        bits = (first << 23) | ((second & 0x8000) << 16) | ((second & 0x7FFF) << 8)
        return struct.unpack("<f", struct.pack("<I", bits))[0]


def align4(value: int) -> int:
    return (value + 3) & ~3


def input_count(convention: int) -> int:
    return (convention & 0x00000F00) >> 8


def link_count(convention: int) -> int:
    return (convention & 0x0000F000) >> 12


def string_count(convention: int) -> int:
    return (convention & 0x00070000) >> 16


def spline_count(convention: int) -> int:
    return (convention & 0x00700000) >> 20


def finite_or_string(value: float) -> float | str:
    if math.isfinite(value):
        return value
    return repr(value)


@dataclass
class KxDocument:
    source: memoryview
    start: int
    end: int
    flags: int
    name: str | None
    song_offset: int
    song_size: int
    sample_offset: int | None
    sample_size: int
    song_bpm_fixed: int
    song_length: int
    roots: list[int]
    classes: list[dict[str, Any]]
    operations: list[dict[str, Any]]
    events: list[dict[str, Any]]
    splines: list[dict[str, Any]]
    sections: dict[str, int]

    @property
    def data(self) -> bytes:
        return bytes(self.source[self.start : self.end])

    @property
    def song_data(self) -> bytes:
        return bytes(self.source[self.song_offset : self.song_offset + self.song_size])

    @property
    def sample_data(self) -> bytes:
        if self.sample_offset is None:
            return b""
        return bytes(self.source[self.sample_offset : self.sample_offset + self.sample_size])

    def summary(self) -> dict[str, Any]:
        class_counts = [0] * len(self.classes)
        animation_bytes = 0
        blob_bytes = 0
        blob_count = 0
        for op in self.operations:
            class_counts[op["class_index"]] += 1
            animation_bytes += op["animation_size"]
            if op["blob_size"]:
                blob_count += 1
                blob_bytes += op["blob_size"]

        classes = []
        for index, cls in enumerate(self.classes):
            classes.append(
                {
                    "index": index,
                    "id": cls["id"],
                    "id_hex": f"0x{cls['id']:04x}",
                    "convention_hex": f"0x{cls['convention']:08x}",
                    "packing": cls["packing"],
                    "operation_count": class_counts[index],
                }
            )

        event_starts = [event["start"] for event in self.events]
        event_ends = [event["end"] for event in self.events]
        spline_channels = sum(len(spline["channels"]) for spline in self.splines)
        spline_keys = sum(
            len(channel)
            for spline in self.splines
            for channel in spline["channels"]
        )
        relative_sections = {
            name: offset - self.start for name, offset in self.sections.items()
        }

        return {
            "format": "werkkzeug3 compact KX player export",
            "size": self.end - self.start,
            "sha256": hashlib.sha256(self.data).hexdigest(),
            "flags": self.flags,
            "has_name": bool(self.flags & 1),
            "has_samples": bool(self.flags & 2),
            "buzz_timing": bool(self.flags & 4),
            "name": self.name,
            "song": {
                "offset": self.song_offset - self.start,
                "size": self.song_size,
                "sha256": hashlib.sha256(self.song_data).hexdigest(),
            },
            "samples": {
                "offset": None
                if self.sample_offset is None
                else self.sample_offset - self.start,
                "size": self.sample_size,
                "sha256": hashlib.sha256(self.sample_data).hexdigest()
                if self.sample_offset is not None
                else None,
            },
            "song_bpm_fixed": self.song_bpm_fixed,
            "song_bpm": self.song_bpm_fixed / 65536.0,
            "song_length": self.song_length,
            "operation_count": len(self.operations),
            "event_count": len(self.events),
            "spline_count": len(self.splines),
            "spline_channel_count": spline_channels,
            "spline_key_count": spline_keys,
            "animation_bytes": animation_bytes,
            "blob_count": blob_count,
            "blob_bytes": blob_bytes,
            "event_start_min": min(event_starts) if event_starts else None,
            "event_end_max": max(event_ends) if event_ends else None,
            "roots": self.roots,
            "sections": relative_sections,
            "classes": classes,
        }


def parse_parameter(cursor: Cursor, packing: str) -> int | float | None:
    kind = packing.lower()
    if kind == "g":
        return cursor.f16()
    if kind == "f":
        return cursor.f24()
    if kind == "e":
        return cursor.s16() / 4096.0
    if kind == "i":
        return cursor.s32()
    if kind == "s":
        return cursor.s16()
    if kind == "b":
        return cursor.u8()
    if kind == "c":
        return cursor.u32()
    if kind == "m":
        return cursor.u24()
    # KDoc::Init ignores non-storage characters in a packing string.
    return None


def parse_animation(cursor: Cursor) -> bytes:
    start = cursor.pos
    while True:
        command = cursor.u8()
        if command >= 0x80:
            cursor.skip(1, "animation store operand")
        else:
            if command >= len(ANIM_PAYLOAD_SIZES):
                raise KxFormatError(
                    f"unknown animation opcode {command:#x} at {cursor.pos - 1:#x}"
                )
            cursor.skip(ANIM_PAYLOAD_SIZES[command], "animation operand")
        if command == KA_END:
            return bytes(cursor.data[start : cursor.pos])


def parse_kx(data: bytes | memoryview, offset: int = 0) -> KxDocument:
    """Parse one document and return its exact self-delimited byte range."""

    source = memoryview(data)
    cursor = Cursor(source, offset)
    start = offset
    sections: dict[str, int] = {"header": start}

    flags = cursor.u32()
    if flags & ~7:
        raise KxFormatError(f"unsupported KX flags {flags:#x}")

    name = None
    if flags & 1:
        raw_name = cursor.take(32, "demo name")
        name = raw_name.split(b"\0", 1)[0].decode("latin-1")

    song_size = cursor.u32()
    song_offset = cursor.skip(align4(song_size), "embedded song")
    sections["song"] = song_offset

    sample_offset: int | None = None
    sample_size = 0
    if flags & 2:
        sample_size = cursor.u32()
        sample_offset = cursor.skip(align4(sample_size), "embedded samples")
        sections["samples"] = sample_offset

    sections["document"] = cursor.pos
    song_bpm_fixed = cursor.u32()
    song_length = cursor.u32()
    operation_count = cursor.compact_short()
    document_spline_count = cursor.compact_short()
    roots = [cursor.compact_short() for _ in range(MAX_OP_ROOT)]

    classes: list[dict[str, Any]] = []
    while True:
        convention = cursor.u32()
        if convention == 0:
            break
        class_id = cursor.u16()
        packing = cursor.cstring("operator packing")
        if len(classes) == 128:
            raise KxFormatError("more than 128 exported operator classes")
        classes.append(
            {"convention": convention, "id": class_id, "packing": packing}
        )
    sections["graph"] = cursor.pos

    operations: list[dict[str, Any]] = []
    for op_index in range(operation_count):
        type_byte = cursor.u8()
        class_index = type_byte & 0x7F
        if class_index >= len(classes):
            raise KxFormatError(
                f"operator {op_index} uses missing class index {class_index}"
            )
        convention = classes[class_index]["convention"]
        if type_byte & 0x80:
            count = 1
        elif convention & OPC_FLEXINPUT:
            count = cursor.u8()
        else:
            count = input_count(convention)

        inputs = []
        for _ in range(count):
            delta = 0 if type_byte & 0x80 else cursor.compact_short()
            target = op_index - 1 - delta
            if target < 0:
                raise KxFormatError(
                    f"operator {op_index} has invalid input delta {delta}"
                )
            inputs.append(target)
        operations.append(
            {
                "class_index": class_index,
                "class_id": classes[class_index]["id"],
                "inputs": inputs,
                "links": [],
                "parameters": [],
                "parameter_offset": None,
                "parameter_size": 0,
                "strings": [],
                "splines": [],
                "blob_size": 0,
                "blob_offset": None,
                "animation_offset": None,
                "animation_size": 0,
            }
        )

    for operation in operations:
        convention = classes[operation["class_index"]]["convention"]
        for _ in range(link_count(convention)):
            encoded = cursor.compact_short()
            target = encoded - 1 if encoded else None
            if target is not None and target >= operation_count:
                raise KxFormatError(f"link target {target} is outside the graph")
            operation["links"].append(target)

    sections["parameters"] = cursor.pos
    for class_index, cls in enumerate(classes):
        convention = cls["convention"]
        for operation in operations:
            if operation["class_index"] != class_index:
                continue
            parameter_start = cursor.pos
            operation["parameter_offset"] = parameter_start - start
            operation["parameters"] = [
                finite_or_string(value) if isinstance(value, float) else value
                for value in (
                    parse_parameter(cursor, packing) for packing in cls["packing"]
                )
            ]
            operation["strings"] = [
                cursor.cstring("operator string")
                for _ in range(string_count(convention))
            ]
            operation["splines"] = [
                (reference - 1 if reference else None)
                for reference in (cursor.u16() for _ in range(spline_count(convention)))
            ]
            if convention & OPC_BLOB:
                operation["blob_size"] = cursor.u32()
            operation["parameter_size"] = cursor.pos - parameter_start

    sections["animations"] = cursor.pos
    for operation in operations:
        operation["animation_offset"] = cursor.pos - start
        animation = parse_animation(cursor)
        operation["animation_size"] = len(animation)

    sections["events"] = cursor.pos
    event_count = cursor.compact_short()
    events: list[dict[str, Any]] = []
    for _ in range(event_count):
        op = cursor.compact_short()
        if op >= operation_count:
            raise KxFormatError(f"event operator {op} is outside the graph")
        start_time = cursor.s32()
        end_time = cursor.s32()
        velocity = cursor.f24()
        modulation = cursor.f24()
        select = cursor.u8()
        scale = [cursor.f24() for _ in range(3)]
        rotate = [cursor.f24() for _ in range(3)]
        translate = [cursor.f24() for _ in range(3)]
        color = cursor.u32()
        spline_ref = cursor.compact_short()
        event_spline = spline_ref - 1 if spline_ref else None
        start_interval = cursor.f16()
        end_interval = cursor.f16()
        event_flags = cursor.u8()
        events.append(
            {
                "operation": op,
                "start": start_time,
                "end": end_time,
                "velocity": finite_or_string(velocity),
                "modulation": finite_or_string(modulation),
                "select": select,
                "scale": [finite_or_string(value) for value in scale],
                "rotate": [finite_or_string(value) for value in rotate],
                "translate": [finite_or_string(value) for value in translate],
                "color": color,
                "spline": event_spline,
                "start_interval": finite_or_string(start_interval),
                "end_interval": finite_or_string(end_interval),
                "flags": event_flags,
            }
        )

    sections["splines"] = cursor.pos
    splines: list[dict[str, Any]] = []
    for _ in range(document_spline_count):
        descriptor = cursor.u8()
        count = descriptor & 7
        interpolation = descriptor >> 3
        if count > 4:
            raise KxFormatError(f"spline has {count} channels")
        channels = []
        for _ in range(count):
            key_count = cursor.compact_short()
            keys = []
            for _ in range(key_count):
                time = cursor.u8() / 255.0
                value = cursor.f16()
                keys.append([time, finite_or_string(value)])
            channels.append(keys)
        splines.append(
            {"interpolation": interpolation, "channels": channels}
        )

    sections["blobs"] = cursor.pos
    for operation in operations:
        size = operation["blob_size"]
        if size:
            operation["blob_offset"] = cursor.pos - start
            cursor.skip(size, "operator blob")

    sections["end"] = cursor.pos
    return KxDocument(
        source=source,
        start=start,
        end=cursor.pos,
        flags=flags,
        name=name,
        song_offset=song_offset,
        song_size=song_size,
        sample_offset=sample_offset,
        sample_size=sample_size,
        song_bpm_fixed=song_bpm_fixed,
        song_length=song_length,
        roots=roots,
        classes=classes,
        operations=operations,
        events=events,
        splines=splines,
        sections=sections,
    )


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", type=Path)
    parser.add_argument("--offset", type=lambda value: int(value, 0), default=0)
    parser.add_argument("--full", action="store_true", help="include graph/event details")
    args = parser.parse_args()

    document = parse_kx(args.input.read_bytes(), args.offset)
    result = document.summary()
    if args.full:
        result["operations"] = document.operations
        result["events"] = document.events
        result["splines"] = document.splines
    print(json.dumps(result, indent=2) + "\n", end="")


if __name__ == "__main__":
    main()
